htslib/sam.h

```diff
@@ -1750,6 +1785,36 @@ int bam_aux_append(bam1_t *b, const char tag[2], char type, int len, const uint8
 HTSLIB_EXPORT
 int bam_aux_del(bam1_t *b, uint8_t *s);
 
+/// Delete an aux field from a BAM record
+/* @param b   The BAM record to update
+   @param s   Pointer to the aux field to delete, as returned by
+              bam_aux_first()/_next()/_get()
+   @return    Pointer to the following aux field, or NULL if none or on error
+
+Identical to @c bam_aux_del() apart from the return value, which is an
+aux iterator suitable for use with @c bam_aux_next()/etc.
+
+Whenever NULL is returned, errno will also be set: ENOENT if the aux field
+deleted was the record's last one; otherwise EINVAL, indicating that the
+BAM record's aux data is corrupt.
+ */
+HTSLIB_EXPORT
+uint8_t *bam_aux_remove(bam1_t *b, uint8_t *s);
+
+/// Delete several aux fields from a BAM record
+/* @param b   The BAM record to update
+   @param s   Pointer to the first aux field to delete, as returned by
+              bam_aux_first()/_next()/_get()
+   @param e   Pointer to the next aux field after the last one to delete,
+              as returned by bam_aux_first()/_next()/_get()
+   @return    Pointer to the following aux field, or NULL if none or on error
+
+When @a e is NULL, removes all fields from @a s onwards. When @s is also NULL,
+no fields are removed.
+ */
+HTSLIB_EXPORT
+uint8_t *bam_aux_remove2(bam1_t *b, uint8_t *s, uint8_t *e);
```

Comment:
What's the use case for `bam_aux_remove2`?

I can see a case for specifying a list of aux tags we wish to remove and it doing the appropriate logic to remove them all within a single pass and avoid potentially O(N^2) complexity on extremely large sets of aux tags.  However unless I'm misunderstanding this, it looks like it's only removing aux tags that are adjacent to each other which is rather tenuous.  I guess the only use I can see is trivial removal of *all* tags.

Although if it's a general list, then it doesn't entirely remove the `samtools view -x` code as that has both keep-list and remove-list functionality.  If we were sufficiently bothered to replicate it here, then the list would need a negation flag to indicate the keep-or-reject status.

Reply:
It's akin to C++ iterators, which have a two-argument `erase()` to remove a consecutive range of items from a sequence. However as you don't directly control the ordering of aux fields, it's not as obviously applicable here. The motivating use case was noted in the first comment above:

> I also have a version with an additional ~~`bam_aux_erase_range()`~~`bam_aux_remove2()` function that deletes several aux fields at once (so could be used to directly reimplement the way James did the deleting in https://github.com/samtools/samtools/pull/516, which probably reduces the number of memmove() calls so is 0.1% more efficient), but not totally sure the slight complication is worth it.

There's a draft of that at jmarshall/samtools@aefbf50cf18fef678beb2b1bf364cda68c2e174c though I think it needs minor alterations. (That commit implements `remove_tag` the obvious way, but attempts the optimised way for `keep_tag`.) The idea is you can implement `keep_tag` by iterating over a run of fields you don't want until you see one that you **do** want to keep, and then nuke the whole run of don't-wants at once with `bam_aux_remove2()`; then repeat that until you're at the end of the aux fields.

Essentially `bam_aux_remove2()` falls out of the implementation so might as well be public. Useful for nuke-all-tagged-fields and as a building block for fancy stuff like `keep_tag`.

Or I'm happy to inline the `aux_remove()` helper back into `bam_aux_remove()` again, so it goes away. :shrug:

Reply:
The problem is that if we're removing X, Y and Z from tags ABCXYDEFZGHI then the first `XY` removal moves DEFZGHI down 2 places, and then the Z moves GHI down 1 place again.   This is because `bam_aux_remove` always moves up to the end (`data + l_data`) and not up to the next point of interest.  Hence GHI has been moved twice.  

The efficient algorithm however is to move DEF down 2 and then GHI down 3.   Hence your reimplementation is not the same complexity as the original.

In general the algorithm is O(N^2) complexity, although in practice it's probably irrelevant as N is usually small.  There may be some cases though, such as the large PacBio tags, where N becomes quite significant.

Reply:
Such is the cost of abstraction. The raw pointer version has the liberty of leaving a gap while it is working; the API version must maintain the invariant that the `bam1_t` has a contiguous vector of valid tag fields between API calls.

Perhaps what would be more useful would be a `bam_aux_remove_if(…)` function that took a callback saying whether to remove a particular aux field, and iterated over the whole list itself within the one API call.

Reply:
Regarding abstraction, that's why I said I felt the `bam_aux_remove2` function to be of questionable benefit.  Perhaps some special corner cases may work, such as removing all bar the first tag, but I don't really see the benefit of extending the API with something so limited.  If we wanted to have an API which always left bam1_t in a consistent manner, then it should be as I said before - a remove function that takes a list of tag names rather than a single one (and perhaps a keep / remove flag to negate it).  Plus we already have this implemented in samtools view.

We could perhaps do it with `bam_aux_remove_if` callbacks, but it seems like a lot of additional faff to use over a basic array argument.

Reply:
I've removed `bam_aux_remove2()`, which was amongst other things a pain to use, and reunified `bam_aux_remove()` accordingly. (See force push to 99954c125b47166945a751b3a0d1fa20284b4cac below.)

I'll add something else more useful for implementing stuff like `keep_tag`/`remove_tag` in its place.